rfcs/convergence/slot-null-rendering.md

```diff
@@ -0,0 +1,261 @@
+      button: 'button',
+      icon: 'i',
+    },
+    button: resolveShorthand(props.button, {
+      children: React.Fragment, // 🚨 getSlots requires children
+    }),
+    icon: resolveShorthand(props.icon),
+  };
+
+  const { slots, slotProps } = getSlots(state, ['input']);
+
+  return (
+    <slots.root {...slotProps.root}>
+      <slots.button {...slotProps.button}>
+        <slots.icon {...slotProps.icon} />
+      </slots.button>
+    </slots.root>
+  );
+}
+```
+
+[Live Example](https://stackblitz.com/edit/react-ts-wwguyp?file=SlotAsParent.tsx)
+
+### Native elements without children
+
+e.g: An Icon that doesn't have children `<i class="icon"/>` (this case is hypothetical, but quite possible, as we cannot verify it in converged components).
+
+## Detailed Design or Proposal
+
+Both options here proposed by the end are equivalent, the only difference is the impact on `ObjectShorthandProps` interface
+
+### Option 1: Make `shorthands` optional
+
+This solves this problem by verifying if the shorthand is `undefined` or not.
+By verifying if shorthand is `undefined` we can opt for null rendering without compromising the cases which native slots don't have children
+
+- `getSlots` should stop verifying null rendering by the existence of the property `children`, instead depending on the existence of the shorthand itself
+
+```tsx
+// react-utilities/src/compose/getSlots.ts
+function getSlots(state, slotNames) {
+  // ...
+  if (state[name] === undefined) {
+    slots[name] = nullRender;
+    continue;
+  }
+  // ...
+}
+```
+
+- Since some slots may have default props, verification must be done on shorthand declaration to ensure that those slots will be null rendered even when presented with default props.
+
+```diff
+export function resolveShorthand(
+  value,
+  defaultProps,
++  {optional = true}
++) {
++  // verify if shorthand is undefined
++  if (value === null || (value === undefined && optional)) {
++    return undefined;
++  }
+}
+```
+
+The problem with this approach is that the shorthand signature now will carry with it an `undefined` value, making overrides a little less elegant:
+
+```ts
+export const CustomAccordionHeader = React.forwardRef<HTMLElement, AccordionHeaderProps>((props, ref) => {
+  const state = useAccordionHeader(props, ref);
+
+  if (state.icon) {
+    state.icon.onClick = () => {
+      /**
+       * this is an override,
+       * if this was a hook than it would need
+       * to be declared before the conditional
+       **/
+    };
+  }
+
+  useAccordionHeaderStyles(state);
+  return renderAccordionHeader(state);
+});
+```
+
+In cases where we simply want to ensure that the slot should be rendered even when `shorthand` is `undefined` than `{optional: false}` should be provided
+
+```ts
+export const useState = (props, ref) => {
+  return {
+    ...props,
+    slot1: resolveShorthand(props.slot1), // undefined
+    slot2: resolveShorthand(props.slot2, {}, { optional: false }), // {}
+  };
+};
+```
+
```

Comment:
Optionally we can modify `resolveShorthand`'s signature, to have second param as config:

```tsx
resolveShorthand(props.slot2, {
  optional: false,
  defaultProps: { "aria-label": "foo" }
});
```

Reply:
hmmm, there's always the possibility that `optional` is actually a property from the slot, and that would generate a conflict hehehe, but yeah, we could assume that slots can't have `optional` as prop

Reply:
Props in this case will be defined under a separate property (`defaultProps`), so they will not collide 😉

Reply:
Totally agree with this idea

Reply:
what @layershifter , adding any kind of positional arguments makes API less readable/hard to maintain/brittle to breaking changes